text/0143-sdk-fail-safe-mode.md

```diff
@@ -0,0 +1,512 @@
+5. It requires infrastructure changes.
+
+### Option B4: [Discarded] Bundling SDK versions at the same time  <a name="option-b4"></a>
+
+> 🚫 **Discard reason:** It has too many significant cons. 🚫
+
+The SDK ships with two different SDK versions. It has a wrapper for the user and then delegates the actual method calls to the duplicated SDK code. If the SDK detects it’s crashing often, it uses the fallback SDK version.
+
+No notes on [crashing scenarios](#crashing-scenarios), because we can discard this option as it has two many significant cons.
+
+#### Pros <a name="option-b4-pros"></a>
+
+1. When the SDK crashes, it can still function with the fallback SDK version.
+
+#### Cons <a name="option-b4-cons"></a>
+
+1. Roughly doubles the size of the SDK.
+2. It requires an extra package.
+3. Only a subset of customers might use this, and only high-quality aware customers might accept the tradeoff of a double-sized SDK. In fact, most high-quality aware customers most likely care about app size and will use the stable release channel.
+
+## C: Knowing When the SDK is Disabled <a name="c-knowing-when-the-sdk-is-disabled"></a>
+
+### Option C1: Failing SDK Endpoint <a name="option-c1"></a>
+
+Add a unique endpoint for sending a simple HTTP request with only the SDK version and a bit of meta-data, such as the DSN, to notify Sentry about failed SDKs. We must keep this logic as simple as possible, and it should hardly ever change to drastically minimize the risk of causing more damage. The HTTP request must not use other parts of the SDK, such as client, hub, or transport. The SDKs must only send this request once. As we can’t have any logic running, such as rate-limiting or client reports, it’s good to have a specific endpoint for this to reduce the potential impact on the rest of the infrastructure.
+
+#### Pros <a name="option-c1-pros"></a>
+
+1. We know when a SDK disables itself.
+
+#### Cons <a name="option-c1-cons"></a>
+
+1. Potential risk of crashing while performing this action.
+2. It requires extra infrastructure.
+3. We don't know why the SDK disabled itself.
+
+### Option C2: SDK Retry Logic <a name="option-c2"></a>
+
+When the SDK either operates in Safe Mode or NoOp Mode, it retries to go back to normal or safe mode after running x times in Safe Mode or NoOp Mode.
+
+#### Pros <a name="option-c2-pros"></a>
+
+1. We don't fly completely blind when the SDK is in NoOp Mode.
+2. We reduce the risk of falsely staying in the NoOp or Safe Mode.
+
+#### Cons <a name="option-c2-cons"></a>
+
+1. The SDK might cause extra crashes, which could confuse users.
+
+### Option C3: Anomaly Detection <a name="option-c3"></a>
+
+The backend detects anomalies in our customers' session data. If there is a significant drop, we can assume that the SDK works in NoOp mode. The logic has to correctly detect debug and staging releases and take sampling into account.
+
```

Comment:
Another idea crossed my mind (idk if we want to add it as an idea or an option here): we could query our customers' google play console/appstore connect via a public API (like an integration) and then look up for our own crashes in the dataset and ingest those into the SDK crash reporting projects and get notified, etc etc. I know it's [possible](https://developers.google.com/play/developer/reporting/reference/rest/v1beta1/vitals.errors.issues) with the google play console API, not sure about the [apple](https://developer.apple.com/documentation/appstoreconnectapi/downloading-analytics-reports) one though.

Reply:
💯 similar to the existing option C3 this could provide a lot of value, even independently of the current RFC. I looked into this a bit for Apple though, and don't think there's a way (at least not a remotely official one). somehow Xcode needs to access this information from the Apple servers, but that's as closed as can get I believe

Reply:
I've worked with the [official App Store Connect API](https://developer.apple.com/documentation/appstoreconnectapi/) quite a lot, and also with [spaceship](https://github.com/fastlane/fastlane/blob/master/spaceship/README.md) which uses unofficial API endpoints used in the App Store Connect web app.

In general, if we need data that is not accessible via public unauthenticated endpoints, the ASC API requires an API key created by the customer, or we need to have access to ASC via an Apple Account (2FA is required).

It's quite complicated to do, but it can be done.

Reply:
maybe a holiday read for me 😅 https://developer.apple.com/documentation/appstoreconnectapi/power-and-performance-metrics-and-logs

Reply:
I have not used the Power and Performance Metrics API before, but it seems to have been available since API version v1.2. 

The last time I have checked crash reports reported to Testflight are still not available in the ASC API, but it can be downloaded with [spaceship.get_beta_feedback](https://github.com/fastlane/fastlane/blob/400f21481aa1b0331b5407e5c4f125ce4b217874/spaceship/lib/spaceship/connect_api/testflight/testflight.rb#L510-L513).